substrate/frame/sassafras/src/tests.rs

```diff
@@ -0,0 +1,742 @@
+		assert_eq!(meta.unsorted_tickets_count, 0);
+		assert_eq!(meta.tickets_count, [0, epoch_length as u32]);
+
+		// Compute and sort the tickets ids (aka tickets scores)
+		let mut expected_ids: Vec<_> = tickets.into_iter().map(|(id, _)| id).collect();
+		expected_ids.sort();
+		expected_ids.truncate(epoch_length as usize);
+
+		// Check if we can fetch next epoch tickets ids (outside-in).
+		let slot = Sassafras::current_slot();
+		assert_eq!(Sassafras::slot_ticket_id(slot + 1).unwrap(), expected_ids[1]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 2).unwrap(), expected_ids[3]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 3).unwrap(), expected_ids[5]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 4).unwrap(), expected_ids[7]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 7).unwrap(), expected_ids[6]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 8).unwrap(), expected_ids[4]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 9).unwrap(), expected_ids[2]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 10).unwrap(), expected_ids[0]);
+		assert!(Sassafras::slot_ticket_id(slot + 11).is_none());
+
+		// Enact epoch change by progressing one more block
+
+		progress_to_block(epoch_length + 1, &pairs[0]).unwrap();
+
+		let meta = TicketsMeta::<Test>::get();
+		assert_eq!(meta.unsorted_tickets_count, 0);
+		assert_eq!(meta.tickets_count, [0, 10]);
+
+		// Check if we can fetch current epoch tickets ids (outside-in).
+		let slot = Sassafras::current_slot();
+		assert_eq!(Sassafras::slot_ticket_id(slot).unwrap(), expected_ids[1]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 1).unwrap(), expected_ids[3]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 2).unwrap(), expected_ids[5]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 3).unwrap(), expected_ids[7]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 6).unwrap(), expected_ids[6]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 7).unwrap(), expected_ids[4]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 8).unwrap(), expected_ids[2]);
+		assert_eq!(Sassafras::slot_ticket_id(slot + 9).unwrap(), expected_ids[0]);
+		assert!(Sassafras::slot_ticket_id(slot + 10).is_none());
+
+		// Enact another epoch change, for which we don't have any ticket
+		progress_to_block(2 * epoch_length + 1, &pairs[0]).unwrap();
+		let meta = TicketsMeta::<Test>::get();
+		assert_eq!(meta.unsorted_tickets_count, 0);
+		assert_eq!(meta.tickets_count, [0, 0]);
+	});
+}
+
+#[test]
+fn block_allowed_to_skip_epochs() {
+	let (pairs, mut ext) = new_test_ext_with_pairs(4, false);
+	let pair = &pairs[0];
+	let start_slot = Slot::from(100);
+	let start_block = 1;
+
+	ext.execute_with(|| {
+		let epoch_length: u64 = <Test as Config>::EpochLength::get();
+
+		initialize_block(start_block, start_slot, Default::default(), pair);
+
+		let tickets = make_ticket_bodies(3, Some(pair));
+		persist_next_epoch_tickets(&tickets);
+
+		let next_random = NextRandomness::<Test>::get();
+
+		// We want to skip 2 epochs in this test.
```

Comment:
You should try skipping an odd number of epochs instead 😈.

Reply:
Done f45ef2f290767577f5ef85a00500373500103c00

The logic still works as when we detect 1+ skipped epoch we just drop everything:
- current ticket data is not required as it is elapsed (as expected)
- next ticket data is not required as it is never started (not expected... but we'll recover)

From this new epoch we begin the AURA like thing, much like what we do for epoch 0.

The validators submits tickets for next epoch as expected, epoch index modulo logic still applies